augur/export_v2.py

```diff
@@ -735,6 +737,77 @@ def _recursively_set_data(node):
             _recursively_set_data(child)
     _recursively_set_data(data_json["tree"])
 
+def is_numeric(n:Any) -> bool:
+    # Typing a number is surprisingly hard in python, and `number.Number`
+    # doesn't work nicely with type hints. See <https://stackoverflow.com/a/73186301>
+    return isinstance(n, (int, float))
+
+def format_number(n: Union[int, float]) -> Union[int, float]:
+    if isinstance(n, int) or n==0:
+        return n
+    # We want to use three sig figs for the fractional part of the float, while
+    # preserving all the data in the integral (integer) part. We leverage the
+    # fact that python floats (incl scientific notation) storing the shortest
+    # decimal string that’s guaranteed to round back to x. Note that this means we
+    # drop trailing zeros, so it's not _quite_ sig figs.
+    integral = int(abs(n))
+    significand = math.floor(math.log10(integral))+1 if integral!=0 else 0
+    return float(f"{n:.{significand+3}g}")
+
+
+class ConfidenceNumeric(TypedDict):
+    confidence: List[Union[int,float]]
```

Comment:
Did you consider doing this as a 2-element tuple instead of a list? I haven't thought through all the impacts of that, but it would codify the "there should always be 2-and-only-2 values" part more explicitly.

Reply:
I didn't, (I think) because the JSON representation is ultimately a list/array. But tuples serialize to arrays in JSON so this should work.

Reply:
oh this will _work_, no doubt, my thought was that writing this as 

```
    confidence: Tuple[Union[int,float],Union[int,float]]
```

more explicitly captures "there should be a list of 2 and only 2 numbers here".

(this is not a big deal or a blocking thing.)

Reply:
Changed in [179cf65](https://github.com/nextstrain/augur/pull/1512/commits/179cf65c4fecac60ff2f191082374a8bc20d2c26)

Interestingly mypy & pyright won't pick up the length check on the `conf` list, so we can't use the following (which is the most similar to the previous code):
```py
    if isinstance(conf, list):
        if len(conf)!=2:
            warn(...)
            return {}
        return {"confidence": tuple(format_number(v) for v in conf)}
```
but we can (and now do) use
```py
    if isinstance(conf, list):
        if len(conf)!=2:
            warn(...)
            return {}
        return {"confidence": (format_number(conf[0]), format_number(conf[1]))}
```

I discovered the type checker is also happy with
```py
    if isinstance(conf, list):
        if len(conf)!=2:
            warn(...)
            return {}
        return {"confidence": (format_number(conf[4]), format_number(conf[5]))}
```
which is guaranteed to be a runtime error and seems a rather large limitation of the type system!

Reply:
_snip_
> seems a rather large limitation of the type system!

**woof** yeah, that's unfortunate, but type hints are only going to take Python so far, I guess — this is a good example of a situation where I think the type information is as much for the edification of future devs as it is for the checker.